### DISCUSSION TITLE
How to avoid repeated consumption in Durable subscription?

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#18170] [General]
My test steps:

(1) Start the producer and produce 50 messages

(2) Start consumers and realize slow consumption with sleep

(3) Ctrl+c aborts the pulsar process

(4) *************************** Because the Consumer has a receiving queue, consumption has not been stopped because of the destruction of the pulsar process. Every time, it is successfully ACKed (is the ACK status information saved locally?)

After restarting the Pulsar and consumers, they will continue to consume at the last location. In this way, repeated consumption occurs,

How to avoid it? thank.

**Comment by Jason918** [comment]
> Every time, it is successfully ACKed (is the ACK status information saved locally?)

Not likely, client `ack` need RPC to be persisted to broker.



> After restarting the Pulsar and consumers, they will continue to consume at the last location. In this way, repeated consumption occurs,
> 
> How to avoid it?

Are you talking about exactly-once message delivery? Normally, we don't need it. The best practice would be implement idempotent message process at consumer side.

If you are talking about minimizing message duplication, it's another story.
